### DISCUSSION TITLE
YOLOX on Kaggle - full custom model training notebook

### DISCUSSION THREAD

**Discussion by rkinas** [Megvii-BaseDetection/YOLOX#957] [Show and tell]
I created and published (probably first) custom model training for Kaggle competitions. Here you can find full pipeline: https://www.kaggle.com/remekkinas/yolox-full-training-pipeline-for-cots-dataset

**Comment by FateScript** [comment]
Good job :)

**Reply by rkinas** [reply]
Thank you! YOLOX is great. Unfortunately I encountered the same problem as described in issue log ... low AP after training. Searching now for solution. I have checked dataset and annotations. I have checked configuration file ... but ... no solution so far. Dataset contains a lot of small objects ... but I am not sure if this is a source of problem.  

Training:
- yolox_s - with pretrained model (Legacy weights model)
- config showed in notebook
- results as well

I know that should be tranined more epoch but look ... on log. Currently I am training on my machine (no resource limited) and gave training more time.

**Reply by rkinas** [reply]
As I can see I have tough dataset .... two models (nano and s) had problem with lerning - after 50-75 epochs IoUs were 0.0. I changed it to M and .... it is learning (after 23 epochs ... for big object - IoUs more then 0.5) ...

**Reply by rkinas** [reply]
From Epoch 17 ... to Epoch  56 ... no changes in IoUs stats .... :(

**Reply by GOATmessi8** [reply]
Hi, I've seen your training log and I think the IoU loss is high.  Maybe you should close the mixup strategy and adjust the "no_aug_epochs" to a small number according to your max epochs.  And we also support rectangle input size now, so you can keep the original 720x1280 resolution. 
Hope this helps

**Reply by rkinas** [reply]
Now we have second part as well: https://www.kaggle.com/remekkinas/yolox-inference-on-kaggle-for-cots
It is time to ... build great model .... :)

**Reply by KeremTurgutlu** [reply]
I think not all rectangle resolutions work, for example with (720,1280) gives error:

```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-39-65fc47b58164> in <module>
----> 1 trainer.model.backbone(inp.cuda())

/usr/local/lib/python3.6/dist-packages/torch/nn/modules/module.py in _call_impl(self, *input, **kwargs)
   1100         if not (self._backward_hooks or self._forward_hooks or self._forward_pre_hooks or _global_backward_hooks
   1101                 or _global_forward_hooks or _global_forward_pre_hooks):
-> 1102             return forward_call(*input, **kwargs)
   1103         # Do not call functions when jit is used
   1104         full_backward_hooks, non_full_backward_hooks = [], []

/trainman-mount/trainman-k8s-storage-5db6d360-c9b6-4daf-8039-164efd47b892/cots/YOLOX/yolox/models/yolo_pafpn.py in forward(self, input)
     97         fpn_out0 = self.lateral_conv0(x0)  # 1024->512/32
     98         f_out0 = self.upsample(fpn_out0)  # 512/16
---> 99         f_out0 = torch.cat([f_out0, x1], 1)  # 512->1024/16
    100         f_out0 = self.C3_p4(f_out0)  # 1024->512/16
    101 

RuntimeError: Sizes of tensors must match except in dimension 1. Expected size 46 but got size 45 for tensor number 1 in the list.
```
however input size (800,1280) works fine - should be fine though since it will be padded:
<img width="785" alt="Screen Shot 2021-12-03 at 8 15 15 PM" src="https://user-images.githubusercontent.com/19826777/144696609-8731abe0-4530-4c0f-b80e-45f49d623e4d.png">

**Reply by rkinas** [reply]
I am training now :) and experiment. We will see :)

**Comment by rkinas** [comment]
I know I know this is to early .... for full happines but ..... YoloX is the best scored notebook currently in this competition https://www.kaggle.com/c/tensorflow-great-barrier-reef/leaderboard I know that everything will change soon but now ... I can fully confirm that full Kaggle training (I update this notebook) and inference pipeline is ready. Hope now people will more often use YoloX in competitions.

**Reply by rkinas** [reply]
Now even better .... LB 0.507 😍😍😍 -> TOP11 and the best public notebook score!!! YoloX .... WoW

**Reply by dsantiago** [reply]
Good job, always making wonderful work on kaggle notebooks

**Reply by rkinas** [reply]
Thank you! :)

**Comment by rkinas** [comment]
Could anybody help me to build in Albumentations (it could be one function eg. Cutout) in training dataloader (hook for Albumentations)?

**Reply by rkinas** [reply]
I managed to write custom Albumentation Data Loader class :) Works fine!

**Reply by jasper-vaart** [reply]
Could you maybe show how you managed to get it working?

I tried adjusting TrainTransform (https://github.com/Megvii-BaseDetection/YOLOX/blob/6880e3999eb5cf83037e1818ee63d589384587bd/yolox/data/data_augment.py) by commenting out the flip and hsv augs; and calling albumentations instead. 
That makes it trainable, however when I visualize the augmentations, the bounding boxes are all wrong.
